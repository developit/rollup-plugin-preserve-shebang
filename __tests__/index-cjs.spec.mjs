import { getRollUpBundle, getFixtureFilePath, getGeneratedContent } from './test-utils.mjs';

describe('shebangPlugin: cjs', () => {
	const format = 'cjs';
	const nodeShebang = '#!/usr/bin/env node';

	it('preseves existing shebang after successful rollup bundle', async () => {
		const inputFilePath = getFixtureFilePath('hello-world-cli.js');

		const bundle = await getRollUpBundle(inputFilePath);
		const generatedCode = await getGeneratedContent(bundle, format);

		expect(generatedCode.startsWith(nodeShebang)).toBe(true);
		expect(generatedCode).toMatchSnapshot();
	});

	it('replaces an existing shebang with a custom shebang', async () => {
		const inputFilePath = getFixtureFilePath('hello-world-cli.js');
		const customShebang = '#!/path/to/custom/interpreter';
		const bundle = await getRollUpBundle(inputFilePath, {
			shebang: customShebang
		});
		const generatedCode = await getGeneratedContent(bundle, format);

		expect(generatedCode.split('\n')[0]).toEqual(customShebang);
		expect(generatedCode.includes(nodeShebang)).toBe(false);
		expect(generatedCode).toMatchSnapshot();
	});

	it('returns unmodified files when no shebang is found', async () => {
		const inputFilePath = getFixtureFilePath('no-shebang.js');
		const bundle = await getRollUpBundle(inputFilePath);
		const generatedCode = await getGeneratedContent(bundle, format);

		expect(generatedCode).toMatchSnapshot();
	});

	it('should retain shebangs defined elsewhere in a file', async () => {
		const inputFilePath = getFixtureFilePath('code-comment-shebang.js');
		const bundle = await getRollUpBundle(inputFilePath);
		const generatedCode = await getGeneratedContent(bundle, format);

		const shebangInstances = generatedCode.split('\n').filter(line => line.includes(nodeShebang));

		expect(shebangInstances.length).toBe(2);
		expect(generatedCode).toMatchSnapshot();
	});

	it('should retain shebangs defined after some whitespace (and remove the leading whitespace)', async () => {
		const inputFilePath = getFixtureFilePath('empty-lines-before-shebang.js');
		const bundle = await getRollUpBundle(inputFilePath);
		const generatedCode = await getGeneratedContent(bundle, format);

		const shebangInstances = generatedCode.split('\n').filter(line => line.includes(nodeShebang));

		expect(generatedCode.split('\n')[0]).toEqual(nodeShebang);
		expect(shebangInstances.length).toBe(3);
		expect(generatedCode).toMatchSnapshot();
	});

	it('should retain comments defined after a shebang', async () => {
		const inputFilePath = getFixtureFilePath('code-comment-after-shebang.js');
		const bundle = await getRollUpBundle(inputFilePath);
		const generatedCode = await getGeneratedContent(bundle, format);

		expect(generatedCode.split('\n')[0]).toEqual('#!/usr/bin/env node // <-- this thing is a shebang. :)');
		expect(generatedCode).toMatchSnapshot();
	});
	
	it('should preserve shebang in files without an EOF character ', async () => {
		const inputFilePath = getFixtureFilePath('cli-without-eof.js');
		const bundle = await getRollUpBundle(inputFilePath);
		const generatedCode = await getGeneratedContent(bundle, format);

		expect(generatedCode.split('\n')[0]).toEqual(nodeShebang);
		expect(generatedCode).toMatchSnapshot();
	});
});

