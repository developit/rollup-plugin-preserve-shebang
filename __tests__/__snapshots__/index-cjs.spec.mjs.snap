// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shebangPlugin: cjs preseves existing shebang after successful rollup bundle 1`] = `
"#!/usr/bin/env node
'use strict';

// eslint-disable-next-line no-console
console.log('Hello, world!');
"
`;

exports[`shebangPlugin: cjs replaces an existing shebang with a custom shebang 1`] = `
"#!/path/to/custom/interpreter
'use strict';

// eslint-disable-next-line no-console
console.log('Hello, world!');
"
`;

exports[`shebangPlugin: cjs returns unmodified files when no shebang is found 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/**
 * Get a unicorn emoji.
 * @returns string
 */
function getUnicornEmoji() {
	return 'ðŸ¦„';
}

exports.getUnicornEmoji = getUnicornEmoji;
"
`;

exports[`shebangPlugin: cjs should preserve shebang in files without an EOF character  1`] = `
"#!/usr/bin/env node
'use strict';

// eslint-disable-next-line no-console
for (let i = 0; i < 10; i++) { console.log('ðŸ¦„');}
"
`;

exports[`shebangPlugin: cjs should retain comments defined after a shebang 1`] = `
"#!/usr/bin/env node // <-- this thing is a shebang. :)
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/**
 * Get a shebang to use in node CLIs. Defaults to #!/usr/bin/env node
 * @returns string
 */
function getNodeSheBang() {
	return '#!/usr/bin/env node';
}

exports.getNodeSheBang = getNodeSheBang;
"
`;

exports[`shebangPlugin: cjs should retain shebangs defined after some whitespace (and remove the leading whitespace) 1`] = `
"#!/usr/bin/env node
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/**
 * Get a shebang to use in node CLIs. Defaults to #!/usr/bin/env node
 * @returns string
 */
function getNodeSheBang() {
	return '#!/usr/bin/env node';
}

exports.getNodeSheBang = getNodeSheBang;
"
`;

exports[`shebangPlugin: cjs should retain shebangs defined elsewhere in a file 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/**
 * Get a shebang to use in node CLIs. Defaults to #!/usr/bin/env node
 * @returns string
 */
function getNodeSheBang() {
	return '#!/usr/bin/env node';
}

exports.getNodeSheBang = getNodeSheBang;
"
`;
